### PR TITLE
Upgrade add-on to Ember.js v5.4.0

### PR DESCRIPTION
**Pull request by MrChocolatine**
### Upgrade add-on to Ember.js v5.4.0 (#255)

- Ember.js v4.8 or above is now required
- Ember CLI v4.8 or above is now required
- Node.js v18 or above is now required

### Partially migrate add-on to TypeScript (#255)

- The add-on is not 100% TypeScript-based yet, but it was regenerated as if it was, with the latest specification and the `--typescript` flag.
  - https://github.com/ember-cli/ember-cli/releases/tag/v4.9.0
  - https://github.com/ember-cli/ember-cli/pull/9972

  which also means it does not rely on `ember-cli-typescript` anymore:
  - https://github.com/ember-cli/ember-cli/pull/10283

#### Added

- `@babel/core@^7.23.2`
- `@ember/string@^3.1.1`
- `@glint/environment-ember-loose@^1.2.1`
- `@glint/template@^1.2.1`
- `@typescript-eslint/eslint-plugin@^6.9.1`
- `@typescript-eslint/parser@^6.9.1`
- `concurrently@^8.2.2`
- `ember-cli-clean-css@^3.0.0`
- `rimraf@^5.0.1`
- `stylelint@^15.11.0`
- `stylelint-config-standard@^34.0.0`
- `stylelint-prettier@^4.0.2`
- `typescript@^5.2.2`
- `@ember/string@^3.1.1` is now a peer dependency
- `ember-source >= 4.0.0` is now a peer dependency

#### Updated

- `ember-auto-import@^2.6.3` is now a dependency, not a dev dependency
- `ember-cli-babel@^8.2.0`
- `ember-cli-htmlbars@^6.3.0`
- `ember-local-storage@^2.0.6`
- `ember-truth-helpers@^4.0.3`
- `@ember/test-helpers@^3.2.0`
- `@embroider/test-setup@^3.0.2`
- `@glimmer/component@^1.1.2`
- `@glimmer/tracking@^1.1.2`
- `ember-auto-import@^2.6.3`
- `ember-cli@~5.4.0`
- `ember-cli-dependency-checker@^3.3.2`
- `ember-page-title@^8.0.0`
- `ember-qunit@^8.0.1`
- `ember-resolver@^11.0.1`
- `ember-source@~5.4.0`
- `ember-template-lint@^5.11.2`
- `ember-try@^3.0.0`
- `eslint@^8.52.0`
- `eslint-config-prettier@^9.0.0`
- `eslint-plugin-ember@^11.11.1`
- `eslint-plugin-n@^16.2.0`
- `eslint-plugin-prettier@^5.0.1`
- `eslint-plugin-qunit@^8.0.1`
- `prettier@^3.0.3`
- `qunit@^2.20.0`
- `webpack@^5.89.0`

#### Removed

- `@babel/core@^7.18.10`
- `@babel/eslint-parser@^7.18.9`
- `@babel/plugin-proposal-decorators@^7.18.10`
- `ember-disable-prototype-extensions@^1.1.3`
- `ember-export-application-global@^2.0.1`
- `npm-run-all@^4.1.5`

---

_Below is just extra text to give more context, but it should not be included in our changelog._

## Actions taken

1. Fresh add-on created using the latest version of Ember.js + flag `--typescript` (using this brings additional changes compared to a simple `ember new MyAddon`)
2. Then re-import everything from our add-on
3. Re-adjust a few things to make it work for our setup

### CODE REVIEW

**Review by ndekeister-us — DISMISSED** [review]
LGTM

**Comment by MrChocolatine** [comment]
Ember.js v5.4.0 has just been released, I will apply the change here while I am at it.
[emberjs/ember.js@v5.4.0 (release)](https://github.com/emberjs/ember.js/releases/tag/v5.4.0)

**Comment by MrChocolatine** [comment]
Calling a last review @ndekeister-us , I dismissed your last approval as I applied another upgrade (v5.3.0 to v5.4.0) after it.

**Comment by MrChocolatine** [comment]
PR initially created the 29 October 2023, after 2 weeks I am going to merge it soon.

File: app/routes/features-list.js
Comment (ndekeister-us) [inline app/routes/features-list.js:1]: I'm not familiar to typescript addons, but why we don't need to export things anymore on `app/` ?

File: index.js
Comment (ndekeister-us) [inline index.js:7]: Is this really needed? On my mind `index.js` file was used by consuming project and it looks strange to me to have a typescript setting here

File: package.json
Comment (ndekeister-us) [inline package.json:110]: For my personal knowledge, is it automatically generated? How you handle updates (like if you update `qunit` package to a new version, you run a specific command to updates `types` too or it need to be done manually?

File: app/routes/features-list.js
Comment (MrChocolatine) [reply app/routes/features-list.js:1]: I deleted this route not due to TS, but for these reasons:

1. In `router.ts`, the `features-list` route is declared
   1. https://github.com/DazzlingFugu/ember-feature-controls/blob/upgrade-to-emberjs-5.3.0/tests/dummy/app/router.ts
   1. https://github.com/DazzlingFugu/ember-feature-controls/blob/upgrade-to-emberjs-5.3.0/addon/route-setup.ts
1. `addon/routes/features-list.js` is a blank Route + because it is declared in the Route => it can be deleted
1. Following (2), `app/routes/features-list.js` can also be deleted.
1. Following (2), `tests/unit/routes/features-list-test.js` can also be deleted.

File: index.js
Comment (MrChocolatine) [reply index.js:7]: https://github.com/ember-cli/ember-cli/blob/v5.4.0/blueprints/addon/files/index.js

File: package.json
Comment (MrChocolatine) [reply package.json:110]: It is from the blueprint:
https://github.com/ember-cli/ember-cli/blob/v5.4.0/blueprints/app/files/package.json

which then gets transformed by the add-on blueprint:
https://github.com/ember-cli/ember-cli/blob/v5.4.0/blueprints/addon/index.js

File: app/routes/features-list.js
Comment (ndekeister-us) [reply app/routes/features-list.js:1]: Yup just tested and you are right, thanks for clarification!

**Review by ndekeister-us — APPROVED** [review]
LGTM

**Pull request merged**
